Mathlib/Algebra/Order/Monovary.lean

```diff
@@ -0,0 +1,428 @@
+@[to_additive] alias ⟨AntivaryOn.of_inv, AntivaryOn.inv⟩ := antivaryOn_inv
+@[to_additive] alias ⟨Monovary.of_inv_left, Antivary.inv_left⟩ := monovary_inv_left
+@[to_additive] alias ⟨Antivary.of_inv_left, Monovary.inv_left⟩ := antivary_inv_left
+@[to_additive] alias ⟨Monovary.of_inv_right, Antivary.inv_right⟩ := monovary_inv_right
+@[to_additive] alias ⟨Antivary.of_inv_right, Monovary.inv_right⟩ := antivary_inv_right
+@[to_additive] alias ⟨Monovary.of_inv, Monovary.inv⟩ := monovary_inv
+@[to_additive] alias ⟨Antivary.of_inv, Antivary.inv⟩ := antivary_inv
+
+@[to_additive] lemma MonovaryOn.mul_left (h₁ : MonovaryOn f₁ g s) (h₂ : MonovaryOn f₂ g s) :
+    MonovaryOn (f₁ * f₂) g s := fun _i hi _j hj hij ↦ mul_le_mul' (h₁ hi hj hij) $ h₂ hi hj hij
+
+@[to_additive] lemma AntivaryOn.mul_left (h₁ : AntivaryOn f₁ g s) (h₂ : AntivaryOn f₂ g s) :
+    AntivaryOn (f₁ * f₂) g s := fun _i hi _j hj hij ↦ mul_le_mul' (h₁ hi hj hij) $ h₂ hi hj hij
+
+@[to_additive] lemma MonovaryOn.div_left (h₁ : MonovaryOn f₁ g s) (h₂ : AntivaryOn f₂ g s) :
+    MonovaryOn (f₁ / f₂) g s := fun _i hi _j hj hij ↦ div_le_div'' (h₁ hi hj hij) $ h₂ hi hj hij
+
+@[to_additive] lemma AntivaryOn.div_left (h₁ : AntivaryOn f₁ g s) (h₂ : MonovaryOn f₂ g s) :
+    AntivaryOn (f₁ / f₂) g s := fun _i hi _j hj hij ↦ div_le_div'' (h₁ hi hj hij) $ h₂ hi hj hij
+
+@[to_additive] lemma MonovaryOn.pow_left (hfg : MonovaryOn f g s) (n : ℕ) :
+    MonovaryOn (f ^ n) g s := fun _i hi _j hj hij ↦ pow_le_pow_of_le_left' (hfg hi hj hij) _
+
+@[to_additive] lemma AntivaryOn.pow_left (hfg : AntivaryOn f g s) (n : ℕ) :
+    AntivaryOn (f ^ n) g s := fun _i hi _j hj hij ↦ pow_le_pow_of_le_left' (hfg hi hj hij) _
+
+@[to_additive]
+lemma Monovary.mul_left (h₁ : Monovary f₁ g) (h₂ : Monovary f₂ g) : Monovary (f₁ * f₂) g :=
+  fun _i _j hij ↦ mul_le_mul' (h₁ hij) $ h₂ hij
+
+@[to_additive]
+lemma Antivary.mul_left (h₁ : Antivary f₁ g) (h₂ : Antivary f₂ g) : Antivary (f₁ * f₂) g :=
+  fun _i _j hij ↦ mul_le_mul' (h₁ hij) $ h₂ hij
+
+@[to_additive]
+lemma Monovary.div_left (h₁ : Monovary f₁ g) (h₂ : Antivary f₂ g) : Monovary (f₁ / f₂) g :=
+  fun _i _j hij ↦ div_le_div'' (h₁ hij) $ h₂ hij
+
+@[to_additive]
+lemma Antivary.div_left (h₁ : Antivary f₁ g) (h₂ : Monovary f₂ g) : Antivary (f₁ / f₂) g :=
+  fun _i _j hij ↦ div_le_div'' (h₁ hij) $ h₂ hij
+
+@[to_additive] lemma Monovary.pow_left (hfg : Monovary f g) (n : ℕ) : Monovary (f ^ n) g :=
+  fun _i _j hij ↦ pow_le_pow_of_le_left' (hfg hij) _
+
+@[to_additive] lemma Antivary.pow_left (hfg : Antivary f g) (n : ℕ) : Antivary (f ^ n) g :=
+  fun _i _j hij ↦ pow_le_pow_of_le_left' (hfg hij) _
+
+end OrderedCommGroup
+
+section LinearOrderedCommGroup
+variable [OrderedCommGroup α] [LinearOrderedCommGroup β] {s : Set ι} {f f₁ f₂ : ι → α}
+  {g g₁ g₂ : ι → β}
+
+@[to_additive] lemma MonovaryOn.mul_right (h₁ : MonovaryOn f g₁ s) (h₂ : MonovaryOn f g₂ s) :
+    MonovaryOn f (g₁ * g₂) s :=
+  fun _i hi _j hj hij ↦ (lt_or_lt_of_mul_lt_mul hij).elim (h₁ hi hj) $ h₂ hi hj
```

Comment:
Does this really need linearity of the order? It feels true without that...

Reply:
`hij` has type `g₁ i * g₂ i < g₁ j * g₂ j`. You can't really do anything with that if you don't have linearity.